### PR TITLE
Coordinated Spawns Handler

### PR DESCRIPTION
**Pull request by Metious**
### Changes made in this pull request

  - Added a new CoordinatedSpawnsHandler that handles and makes Coordinated (Vector3) spawns possible using classIDs or TechTypes.
  - Added a new virtual property to Spawnable.cs for Coordinated spawns.

all of the prefabs made using this new handler work fully Asynchronously in preparation of the upcoming SN experimental branch and BZ.

### CODE REVIEW

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+using UnityEngine;
+
+namespace SMLHelper.V2.Handlers
+{
+    /// <summary>
+    /// a Handler that handles and registers Coordinated (<see cref="Vector3"/> spawns).
+    /// </summary>
+    public class CoordinatedSpawnsHandler : ICoordinatedSpawnHandler
+    {
+        public static ICoordinatedSpawnHandler Main { get; } = new CoordinatedSpawnsHandler();
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:13]: Missing XML Documentation

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            // Hide Constructor
+        }
+
+        #region Interface Implementations
+        /// <summary>
+        /// Registers Multiple Coordinated spawns for one single passed TechType
+        /// </summary>
+        /// <param name="techTypeToSpawn">The TechType to spawn</param>
+        /// <param name="coordinatesToSpawnTo">the coordinates the <see cref="techTypeToSpawn"/> should spawn to</param>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:25]: cref="techTypeToSpawn" should be cref="TechType"

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                spawnInfos.Add(new SpawnInfo(techTypeToSpawn, coordinate));
+            }
+            LargeWorldStreamerPatcher.spawnInfos.AddRange(spawnInfos);
+        }
+        
+        /// <summary>
+        /// Registers a Coordinated Spawn
+        /// </summary>
+        /// <param name="spawnInfo">the SpawnInfo to spawn</param>
+        void ICoordinatedSpawnHandler.RegisterCoordinatedSpawn(SpawnInfo spawnInfo) =>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:40]: Please use block body for methods

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            LargeWorldStreamerPatcher.spawnInfos.Add(spawnInfo);
+
+        /// <summary>
+        /// registers Many Coordinated Spawns.
+        /// </summary>
+        /// <param name="spawnInfos">The SpawnInfo to spawn.</param>
+        void ICoordinatedSpawnHandler.RegisterCoordinatedSpawns(List<SpawnInfo> spawnInfos) =>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:47]: Please use block body for methods

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            LargeWorldStreamerPatcher.spawnInfos.AddRange(spawnInfos);
+        #endregion
+
+        #region Static Methods
+        /// <summary>
+        /// Registers a Coordinated Spawn
+        /// </summary>
+        /// <param name="spawnInfo">the SpawnInfo to spawn</param>
+        public static void RegisterCoordinatedSpawn(SpawnInfo spawnInfo) =>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:56]: Please use block body for methods

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            Main.RegisterCoordinatedSpawn(spawnInfo);
+
+        /// <summary>
+        /// registers Many Coordinated Spawns.
+        /// </summary>
+        /// <param name="spawnInfos">The SpawnInfo to spawn.</param>
+        public static void RegisterCoordinatedSpawns(List<SpawnInfo> spawnInfos) =>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:63]: Please use block body for methods

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            Main.RegisterCoordinatedSpawns(spawnInfos);
+
+        /// <summary>
+        /// Registers Multiple Coordinated spawns for one single passed TechType
+        /// </summary>
+        /// <param name="techTypeToSpawn">The TechType to spawn</param>
+        /// <param name="coordinatesToSpawnTo">the coordinates the <see cref="techTypeToSpawn"/> should spawn to</param>
+        public static void RegisterCoordinatedSpawnsForOneTechType(TechType techTypeToSpawn, List<Vector3> coordinatesToSpawnTo) =>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:71]: Please use block body for methods

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        /// </summary>
+        /// <param name="spawnInfos">The SpawnInfo to spawn.</param>
+        public static void RegisterCoordinatedSpawns(List<SpawnInfo> spawnInfos) =>
+            Main.RegisterCoordinatedSpawns(spawnInfos);
+
+        /// <summary>
+        /// Registers Multiple Coordinated spawns for one single passed TechType
+        /// </summary>
+        /// <param name="techTypeToSpawn">The TechType to spawn</param>
+        /// <param name="coordinatesToSpawnTo">the coordinates the <see cref="techTypeToSpawn"/> should spawn to</param>
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:70]: cref="techTypeToSpawn" should be cref="TechType"

File: SMLHelper/Interfaces/ICoordinatedSpawnHandler.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+using System.Collections.Generic;
+using SMLHelper.V2.Handlers;
+using UnityEngine;
+
+namespace SMLHelper.V2.Interfaces
+{
+    public interface ICoordinatedSpawnHandler
```
Comment (MrPurple6411) [inline SMLHelper/Interfaces/ICoordinatedSpawnHandler.cs:7]: Missing ALL XML Documentation

File: SMLHelper/MonoBehaviours/EntitySpawner.cs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        void Start() => StartCoroutine(SpawnAsync());
+
+        IEnumerator SpawnAsync()
+        {
+            var task = new TaskResult<GameObject>();
+            yield return GetPrefabAsync(task);
+
+            var obj = UWE.Utils.InstantiateDeactivated(task.Get(), spawnInfo.spawnPosition, Quaternion.identity);
+
```
Comment (MrPurple6411) [inline SMLHelper/MonoBehaviours/EntitySpawner.cs:20]: Add a null check with error logging on the result just incase the get prefab fails to get a prefab

File: SMLHelper/MonoBehaviours/EntitySpawner.cs [hunk 11]
```diff
@@ -0,0 +1,64 @@
+using System.Collections;
+using SMLHelper.V2.Handlers;
+using UnityEngine;
+using UWE;
+
+namespace SMLHelper.V2.MonoBehaviours
+{
+    public class EntitySpawner : MonoBehaviour
```
Comment (MrPurple6411) [inline SMLHelper/MonoBehaviours/EntitySpawner.cs:8]: XML Docs

File: SMLHelper/MonoBehaviours/EntitySpawner.cs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        public SpawnInfo spawnInfo;
```
Comment (MrPurple6411) [inline SMLHelper/MonoBehaviours/EntitySpawner.cs:10]: XML Docs

File: SMLHelper/Handlers/CoordinatedSpawnsHandler.cs [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    
+    #region SpawnInfo
+    /// <summary>
+    /// a basic class that provides enough info for the <see cref="CoordinatedSpawnsHandler"/> System to function.
+    /// </summary>
+    public class SpawnInfo
+    {
+        public TechType techType;
+        public string classId;
+        public Vector3 spawnPosition;
```
Comment (MrPurple6411) [inline SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:100]: Do these 3 need to be public or can they be internal?   If public then they need XML Docs

Comment (MrPurple6411) [reply SMLHelper/Handlers/CoordinatedSpawnsHandler.cs:100]: tbh I think these should be internal as they have constructors to deal with the assignment.

Comment (Metious) [reply SMLHelper/Handlers/CoordinatedSpawnsHandler.cs]: was pushing a commit for them when u sent this comment :)

**Comment by Metious** [comment]
Everything has been tested and seems to perform as intended.

**Review by MrPurple6411 — APPROVED** [review]

**Review by PrimeSonic — COMMENTED** [review]
Merge whenever you're ready

**Pull request merged**